Add unit tests for `TimeSnapshot` command in `GeneralDiagnosticsCluster`

LyudmilaKostanyan opened the pull request:
#### Summary

This PR adds unit tests for the `TimeSnapshot` command in the General Diagnostics cluster and refactors existing tests to use helper functions from `ValidateGlobalAttributes.h`.

**What was added:**

- `TimeSnapshotCommandTest`: Verifies that `TimeSnapshot` command returns valid system time without POSIX time in basic configuration

- `TimeSnapshotCommandWithPosixTimeTest`: Validates that POSIX time is included in the response when the feature is enabled

- `TimeSnapshotResponseValues`: Ensures time values are monotonic and never go backwards across multiple command invocations

**What was refactored:**

- `AttributesTest`: Simplified command and attribute validation by using helper functions
- Fixed typos

#### Related issues

Main issue: #41127

#### Testing

This PR only adds new unit tests and refactors existing ones. No changes were made to production code.

LyudmilaKostanyan committed:
Add unit tests for `TimeSnapshot` command in `GeneralDiagnosticsCluster`

- Introduced a mock CommandHandler and a NullProvider for testing.
- Implemented tests for TimeSnapshot command, including scenarios with and without POSIX time support.
- Verified response values and ensured correct handling of command invocations.
- Added helper methods for setting up TLV readers and buffers for command testing.

andy31415 commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -247,4 +352,133 @@ TEST_F(TestGeneralDiagnosticsCluster, AttributesTest)
+    DataModel::InvokeRequest invokeRequest;
+    invokeRequest.path = commandPath;
+
+    uint8_t tlvBuffer[kEmptyCommandTLVBufferSize];
```
I think we should work on some infrastructure on writing these tests somehow cleaner/smaller. Allocating buffers and calling helpers seems painful.

We have some (but maybe not ideal) in https://github.com/project-chip/connectedhomeip/blob/master/src/app/data-model-provider/tests/ReadTesting.h for example and the same for Write. Maybe we should have something for commands.

LyudmilaKostanyan committed:
Refactor `TestGeneralDiagnosticsCluster` to improve command invocation testing

- Moved command invocation logic into a helper method  for better code reuse and clarity.

LyudmilaKostanyan commented:
/gemini review

Elen777300 commented on src/app/clusters/testing/CommandTesting.h:
```diff
@@ -0,0 +1,192 @@
+    // Invoke a command with no fields (empty TLV structure)
+    // Use for commands that take no parameters
+    template <typename ClusterType>
+    std::optional<DataModel::ActionReturnStatus> InvokeEmpty(ClusterType & cluster)
```
```suggestion
    template <typename ClusterType, typename... Args>
std::optional<DataModel::ActionReturnStatus> Invoke(ClusterType & cluster, const Args &... args)
```
Can we combine the two functions in a one, what do you think ?

LyudmilaKostanyan replied:
Done

Elen777300 commented:
I think it is better to separate this into two PRs, because this includes changes in other unit test.
It will be easier to review too

LyudmilaKostanyan committed:
Handle status responses in `ClusterTester` when `InvokeCommand` returns `nullopt`

When InvokeCommand returns nullopt, check for both data responses and
status responses in the mock handler. Previously only data responses
were checked. Now if a status response exists, use the last status
from the handler.

LyudmilaKostanyan commented on src/app/clusters/testing/ClusterTester.h:
```diff
@@ -175,6 +175,12 @@ class ClusterTester
+        // If InvokeCommand returned nullopt but there's a response, treat it as success
+        if (!result.status.has_value() && mHandler.HasResponse())
+        {
+            result.status = app::DataModel::ActionReturnStatus(CHIP_NO_ERROR);
```
Done

LyudmilaKostanyan commented:
/gemini review

LyudmilaKostanyan commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -247,4 +274,69 @@ TEST_F(TestGeneralDiagnosticsCluster, AttributesTest)
+    InvokeTimeSnapshotAndGetResponse(cluster, response);
+
+    // Verify that systemTimeMs is present
+    EXPECT_GE(response.systemTimeMs, 0u);
```
Done

LyudmilaKostanyan committed:
Handle status responses and unexpected states in ClusterTester

When InvokeCommand returns nullopt, check for both data responses
and status responses. If a status response exists, use the last
status from the handler. Also handle the unexpected case where
neither response nor status is provided by setting an error status
and logging a warning.

LyudmilaKostanyan commented:
/gemini review

Elen777300 commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -129,7 +151,7 @@ TEST_F(TestGeneralDiagnosticsCluster, AttributesTest)
                   CHIP_NO_ERROR);
 
-        ASSERT_TRUE(Testing::EqualAttributeSets(attributesBuilder.TakeBuffer(), expectedBuilder.TakeBuffer()));
+        ASSERT_TRUE(chip::Testing::EqualAttributeSets(attributesBuilder.TakeBuffer(), expectedBuilder.TakeBuffer()));
```
https://github.com/project-chip/connectedhomeip/pull/41291/files#r2435994388
https://github.com/project-chip/connectedhomeip/pull/41291/files#r2445449769

LyudmilaKostanyan commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -57,10 +62,35 @@ class ScopedDiagnosticsProvider
+        ASSERT_TRUE(startupError == CHIP_NO_ERROR || startupError == CHIP_ERROR_ALREADY_INITIALIZED);
+
+        GeneralDiagnostics::Commands::TimeSnapshot::Type request{};
+        auto result = tester.Invoke<GeneralDiagnostics::Commands::TimeSnapshotResponse::DecodableType,
```
Done

LyudmilaKostanyan commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -57,10 +62,35 @@ class ScopedDiagnosticsProvider
+    static void InvokeTimeSnapshotAndGetResponse(ClusterType & cluster,
+                                                 GeneralDiagnostics::Commands::TimeSnapshotResponse::DecodableType & response)
+    {
+        ClusterTester tester(cluster);
```
Done

LyudmilaKostanyan commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -57,10 +60,35 @@ class ScopedDiagnosticsProvider
+
+        GeneralDiagnostics::Commands::TimeSnapshot::Type request{};
+        auto result = tester.Invoke<GeneralDiagnostics::Commands::TimeSnapshotResponse::DecodableType,
+                                    GeneralDiagnostics::Commands::TimeSnapshot::Type>(
```
Done

LyudmilaKostanyan commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -57,10 +60,40 @@ class ScopedDiagnosticsProvider
+        ASSERT_TRUE(result.status->IsSuccess()); // NOLINT(bugprone-unchecked-optional-access)
+        ASSERT_TRUE(result.response.has_value());
+        response = result.response.value(); // NOLINT(bugprone-unchecked-optional-access)
+    }
```
Done

andy31415 commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -247,4 +279,63 @@ TEST_F(TestGeneralDiagnosticsCluster, AttributesTest)
+    GeneralDiagnosticsCluster cluster(optionalAttributeSet);
+
+    // Invoke TimeSnapshot command and get response
+    SetCluster(cluster);
```
This `SetCluster` usage seems off .... How about having this code read like:

```cpp
    ClusterTester tester(cluster);
    ASSERT_SUCCESS(cluster.Startup(tester.GetServerClusterContext());
    GeneralDiagnostics::Commands::TimeSnapshotResponse::DecodableType response;
    ASSERT_SUCCESS(InvokeTimeSnapshotAndGetResponse(cluster, response));
```

With required changes to the invoke method.

LyudmilaKostanyan replied:
Done

andy31415 commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -108,9 +140,9 @@ TEST_F(TestGeneralDiagnosticsCluster, AttributesTest)
 
         // Check required generated commands are present
-        ReadOnlyBufferBuilder<chip::CommandId> generatedCommandsBuilder;
+        ReadOnlyBufferBuilder<CommandId> generatedCommandsBuilder;
```
we should probably switch these to use the helpers in https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/testing/ValidateGlobalAttributes.h#L65

LyudmilaKostanyan committed:
Refactor `TestGeneralDiagnosticsCluster` to improve command invocation and response handling

- Moved the `InvokeTimeSnapshotAndGetResponse` method outside the test struct for better organization and clarity.
- Simplified the command invocation process by ensuring `ClusterTester` is instantiated within each test case.
- Enhanced readability by removing redundant code and ensuring consistent error handling during command execution.

LyudmilaKostanyan commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -57,10 +62,32 @@ class ScopedDiagnosticsProvider
+    static void TearDownTestSuite() { Platform::MemoryShutdown(); }
+
+    // Helper method to invoke TimeSnapshot command and decode response
+    void InvokeTimeSnapshotAndGetResponse(GeneralDiagnosticsCluster & cluster,
```
Done

LyudmilaKostanyan commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -57,10 +62,32 @@ class ScopedDiagnosticsProvider
+    {
+        ClusterTester tester(cluster);
+        CHIP_ERROR startupError = cluster.Startup(tester.GetServerClusterContext());
+        ASSERT_TRUE(startupError == CHIP_NO_ERROR || startupError == CHIP_ERROR_ALREADY_INITIALIZED);
```
Done

LyudmilaKostanyan commented:
/gemini review

LyudmilaKostanyan committed:
Refactor `InvokeTimeSnapshotAndGetResponse` to use `ClusterTester` directly in tests

- Updated the `InvokeTimeSnapshotAndGetResponse` function to accept `ClusterTester` instead of `GeneralDiagnosticsCluster` for improved clarity and consistency in test cases.
- Adjusted all relevant test cases to reflect this change, ensuring proper command invocation and response handling.

LyudmilaKostanyan committed:
Enhance error handling in `InvokeTimeSnapshotAndGetResponse` for improved robustness

- Updated the InvokeTimeSnapshotAndGetResponse function to return appropriate error codes for various failure scenarios, ensuring better error handling during command execution.
- Adjusted test cases to assert the return value of the function, enhancing the reliability of the tests and ensuring that errors are properly captured and reported.

LyudmilaKostanyan committed:
Fix test helper to properly propagate errors

Changed InvokeTimeSnapshotAndGetResponse() to return CHIP_ERROR instead
of void. ASSERT macros inside void helper functions only return from the
helper, not from the test itself, which can lead to tests continuing with
uninitialized data and undefined behavior.

LyudmilaKostanyan commented on src/app/clusters/general-diagnostics-server/tests/TestGeneralDiagnosticsCluster.cpp:
```diff
@@ -57,10 +62,38 @@ class ScopedDiagnosticsProvider
+        return CHIP_ERROR_INTERNAL;
+    }
+    response = result.response.value();
+    return CHIP_NO_ERROR;
```
Done

LyudmilaKostanyan commented:
/gemini review

LyudmilaKostanyan commented:
/gemini review
